contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -0,0 +1,286 @@
+
+        // Avoid execution instead of reverting in case a batch includes an already executed request
+        if (signerMatch && alive) {
+            // Nonce should be used before the call to prevent reusing by reentrancy
+            uint256 currentNonce = _useNonce(signer);
+
+            (success, ) = request.to.call{gas: request.gas, value: request.value}(
+                abi.encodePacked(request.data, request.from)
+            );
+
+            _checkForwardedGas(request);
+
+            emit ExecutedForwardRequest(signer, currentNonce, success);
+        }
+    }
+
+    /**
+     * @dev Checks if the requested gas was correctly forwarded to the callee.
+     *
+     * As a consequence of https://eips.ethereum.org/EIPS/eip-150[EIP-150]:
+     * - At most `gasleft() - floor(gasleft() / 64)` is forwarded to the callee.
+     * - At least `floor(gasleft() / 64)` is kept in the caller.
+     *
+     * It reverts consuming all the available gas if the forwarded gas is not the requested gas.
+     *
+     * IMPORTANT: This function should be called exactly the end of the forwarded call. Any gas consumed
+     * in between will make room for bypassing this check.
+     */
+    function _checkForwardedGas(ForwardRequestData calldata request) private view {
+        // To avoid insufficient gas griefing attacks, as referenced in https://ronan.eth.limo/blog/ethereum-gas-dangers/
+        //
+        // A malicious relayer can attempt to shrink the gas forwarded so that the underlying call reverts out-of-gas
+        // and the top-level call still passes, so in order to make sure that the subcall received the requested gas,
+        // the define this model and adding a check:
+        //
+        // Let X be the gas available before the subcall, such that the subcall gets X * 63 / 64.
+        // We can't know X after CALL dynamic costs, but we want it to be such that X * 63 / 64 >= req.gas.
+        // Let Y be the gas used in the subcall gasleft() measured immediately after the subcall will be gasleft() = X - Y.
+        // If the subcall ran out of gas, then Y = X * 63 / 64 and gasleft() = X - Y = X / 64.
+        // Then we restrict the model by checking if req.gas / 63 > gasleft(), which is true is true if and only if
+        // req.gas / 63 > X / 64, or equivalently req.gas > X * 63 / 64.
+        //
+        // This means that if the subcall runs out of gas we are able to detect that insufficient gas was passed.
+        // We will now also see that req.gas / 63 > gasleft() implies that req.gas >= X * 63 / 64.
+        // The contract guarantees Y <= req.gas, thus gasleft() = X - Y >= X - req.gas.
+        // -    req.gas / 63 > gasleft()
+        // -    req.gas / 63 >= X - req.gas
+        // -    req.gas >= X * 63 / 64
+        //
+        // In other words if req.gas < X * 63 / 64 then req.gas / 63 <= gasleft(), thus if the relayer behaves honestly
+        // the relay does not revert.
+        if (gasleft() < request.gas / 63) {
```

Comment:
We are missing a piece here:

```suggestion
        if (gasleft() < request.gas / 63 && !success) {
```

This is related to where the proof above says "If the subcall ran out of gas".

We need to add success as an argument.

Reply:
I am not sure about this, I think the point of the proof was actually proving we don't need to check the success. If it fails, the check will catch it, but if it succeeds there's no way the relayer manipulated the gas provided anyway.

Reply:
I got this idea at some point but I just reviewed the proof and you're correct.